simulator/infra/cache/cache_tag_array.h

```diff
@@ -1,82 +1,191 @@
+        // as the first value in a pair
+        std::pair<Key, std::size_t> update( const Key& key)
+        {
+            std::pair<Key, std::size_t> return_value;
+
+            auto data_it = data.find( key);
+            auto lru_it  = lru_hash.find( key);
+            assert( ( data_it == data.end()) == ( lru_it == lru_hash.end()));
+
+            if ( data_it == data.end())
+                return_value = allocate( key);
+            else
+            {
+                lru_list.splice( lru_list.begin(), lru_list, lru_it->second);
+                return_value = std::pair<Key, std::size_t>( key, data_it->second);
+            }
+
+            return return_value;    
+        }
+
+
+    private:
+        // if the number of elements is equal to CAPACITY,
+        // a key of the least recently used elements is returned 
+        // as the first value in a pair
+
+        // if the number of elements is less than CAPACITY,
+        // a passed key is returned as the first value in a pair
+        std::pair<Key, std::size_t> allocate( const Key& key)
+        {
+            std::size_t tag_value = 0;
+            Key key_value = key;
+
+            if ( number_of_elements == CAPACITY)
+            {
+                // Delete least recently used element and save its key and value
+                const auto& lru_elem = lru_list.back();
+                
+                tag_value = data.find( lru_elem)->second;
+                key_value = lru_elem;
+                
+                lru_hash.erase( lru_elem);
+                data.erase( lru_elem);
+                lru_list.pop_back();
+            }
+            else
+            {
+                tag_value = number_of_elements;
+                number_of_elements++;
+            }
+
+            data.emplace( key, tag_value);
+            auto ptr = lru_list.insert( lru_list.begin(), key);
+            lru_hash.emplace( key, ptr);
+
+            return std::pair<Key, std::size_t>( key_value, tag_value);
+        }
+
+    private:
+        std::unordered_map<Key, std::size_t> data{};
+
+        std::list<Key> lru_list{};
+        std::unordered_map<Key, typename std::list<Key>::const_iterator> lru_hash{};
+
+        std::size_t number_of_elements = 0u;
+        std::size_t CAPACITY;
 };
 
-class CacheTagArrayCheck : private Log
+
+
+class LRUModule
 {
-protected:
-    CacheTagArrayCheck( uint32 size_in_bytes,
-                        uint32 ways,
-                        uint32 line_size,
-                        uint32 addr_size_in_bits);
-public:
-        const uint32 size_in_bytes;
-        const uint32 ways;
-        const uint32 line_size;
-        const uint32 addr_size_in_bits;
+    public:
+        LRUModule( std::size_t number_of_sets, std::size_t number_of_ways)
+            : lru_info( number_of_sets, LRUCacheInfo<Addr>( number_of_ways))
+        { }
+
+        std::pair<Addr, uint32> update( uint32 num_set, Addr addr) 
```

Comment:
There should be two methods:
* Touch (when an existing element is accessed)
* Update (when an new element is added)